### PR TITLE
Replace TT Lib Transpose with TTNN Transpose

### PR DESCRIPTION
**Pull request by ntarafdar**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/10536)

### Problem description
Migrating TTLib transpose to TTNN

### What's changed
Replaced all TT Lib C++ code into TTNN

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

FYI: will not merge until post-commit CI passes, all t3k pipelines, model perf, device perf and nightly

### CODE REVIEW

**Review by tt-rkim — APPROVED** [review]

**Review by TT-BrianLiu — COMMENTED** [review]
Can we just delete the old code in one shot? It should be straightforward

**Comment by tarafdarTT** [comment]
> Can we just delete the old code in one shot? It should be straightforward

I can but then I'd need to run all pipelines that use it (t3k, models etc. )

**Comment by tarafdarTT** [comment]
> > Can we just delete the old code in one shot? It should be straightforward
> 
> I can but then I'd need to run all pipelines that use it (t3k, models etc. )

@TT-BrianLiu you're right though, even with the extra pipelines, if there is an intermediate state with two versions, there's a possibility of stale updates getting pushed through. I'll delete the old code to make sure there's no replication.

File: models/experimental/roberta/tt/roberta_intermediate.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def __init__(self, config, state_dict, base_address, device, fall_back_to_torch_
         )
 
     def linear(self, x, weight, bias):
-        weight = tt_lib.tensor.transpose(weight, -2, -1)
+        weight = ttnn.transposeweight, -2, -1)
```
Comment (ayerofieiev-tt) [inline models/experimental/roberta/tt/roberta_intermediate.py:40]:
```suggestion
        weight = ttnn.transpose(weight, -2, -1)
```

File: models/experimental/roberta/tt/roberta_for_token_classification.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def __init__(self, config, state_dict, base_address, device, reference_model):
         self.classifier_bias = torch2tt_tensor(state_dict[f"classifier.bias"], self.device)
 
     def linear(self, x, weight, bias):
-        weight = tt_lib.tensor.transpose(weight, -2, -1)
+        weight = ttnn.transposeweight, -2, -1)
```
Comment (ayerofieiev-tt) [inline models/experimental/roberta/tt/roberta_for_token_classification.py:55]:
```suggestion
        weight = ttnn.transpose(weight, -2, -1)
```

File: models/experimental/roberta/tt/roberta_for_question_answering.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def __init__(
         self.qa_outputs_bias = torch2tt_tensor(state_dict[f"qa_outputs.bias"], self.device)
 
     def linear(self, x, weight, bias):
-        weight = tt_lib.tensor.transpose(weight, -2, -1)
+        weight = ttnn.transposeweight, -2, -1)
```
Comment (ayerofieiev-tt) [inline models/experimental/roberta/tt/roberta_for_question_answering.py:61]:
```suggestion
        weight = ttnn.transpose(weight, -2, -1)
```

File: models/experimental/nanogpt/tt/nanogpt_mlp.py [hunk 4]
```diff
@@ -29,8 +29,8 @@ def __init__(self, base_address, config, device, tt_cache_path, dtype):
             tt_cache_path + base_address + ".c_proj.bias" + str(dtype) + ".bin"
         )
 
-        self.tt_weight_c_fc = tt_lib.tensor.transpose(self.tt_weight_c_fc, -2, -1)
-        self.tt_weight_c_proj = tt_lib.tensor.transpose(self.tt_weight_c_proj, -2, -1)
+        self.tt_weight_c_fc = ttnn.transpose(self.tt_weight_c_fc, -2, -1)
+        self.tt_weight_c_proj = ttnn.transpos(eself.tt_weight_c_proj, -2, -1)
```
Comment (ayerofieiev-tt) [inline models/experimental/nanogpt/tt/nanogpt_mlp.py:33]:
```suggestion
        self.tt_weight_c_proj = ttnn.transpose(self.tt_weight_c_proj, -2, -1)
```

File: models/experimental/roberta/tt/roberta_lm_head.py [hunk 5]
```diff
@@ -58,7 +58,7 @@ def __init__(
             self.decoder.bias.data = state_dict[f"{base_address}.decoder.bias"]
 
     def linear(self, x, weight, bias):
-        weight = tt_lib.tensor.transpose(weight, -2, -1)
+        weight = ttnn.transposeweight, -2, -1)
```
Comment (ayerofieiev-tt) [inline models/experimental/roberta/tt/roberta_lm_head.py:61]:
```suggestion
        weight = ttnn.transpose(weight, -2, -1)
```

File: models/experimental/roberta/tt/roberta_pooler.py [hunk 6]
```diff
@@ -41,7 +41,7 @@ def __init__(
         )
 
     def linear(self, x, weight, bias):
-        weight = tt_lib.tensor.transpose(weight, -2, -1)
+        weight = ttnn.transposeweight, -2, -1)
```
Comment (ayerofieiev-tt) [inline models/experimental/roberta/tt/roberta_pooler.py:44]:
```suggestion
        weight = ttnn.transpose(weight, -2, -1)
```

File: ttnn/cpp/ttnn/operations/data_movement/transpose/device/cn_multi_core/transpose_cn_op_multi_core.cpp [hunk 7]
```diff
@@ -2,7 +2,7 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#include "ttnn/deprecated/tt_dnn/op_library/transpose/transpose_op.hpp"
+#include "../transpose_op.hpp"
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/transpose/device/cn_multi_core/transpose_cn_op_multi_core.cpp:5]: Absolute path is better

Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/transpose/device/cn_multi_core/transpose_cn_op_multi_core.cpp:5]: Please don't use relative includes like this

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/data_movement/transpose/device/cn_multi_core/transpose_cn_op_multi_core.cpp:5]:
```
#include "ttnn/operations/......"
```

File: ttnn/cpp/ttnn/operations/data_movement/transpose/device/hc_multi_core/transpose_hc_op_multi_core.cpp [hunk 8]
```diff
@@ -2,7 +2,7 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-#include "ttnn/deprecated/tt_dnn/op_library/transpose/transpose_op.hpp"
+#include "../transpose_op.hpp"
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/transpose/device/hc_multi_core/transpose_hc_op_multi_core.cpp:5]: replace relative include

File: ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp [hunk 9]
```diff
@@ -6,24 +6,95 @@
...
+#include "device/transpose_op.hpp"
+#include "ttnn/operations/data_movement/permute/permute.hpp"
 
 
 namespace ttnn {
 namespace operations::data_movement {
 
 constexpr uint32_t TransposeDefaultQueueID = 0;
 
+inline Tensor transpose_(const Tensor &a, TransposeOpDim transpose_dim, const MemoryConfig& output_mem_config) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp:18]: is this an in-plaice version?

File: ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_op.cpp [hunk 10]
```diff
@@ -122,15 +121,15 @@ operation::ProgramWithCallbacks Transpose::create_program(const std::vector<Tens
...
-                return transpose_wh_multi_core(input_tensor, output_tensor);
+                return ttnn::operations::data_movement::transpose_wh_multi_core(input_tensor, output_tensor);
             }
             break;
         case TransposeOpParallelizationStrategy::MULTI_CORE_HC:
-            return transpose_hc_multi_core(input_tensor, output_tensor);
+            return ttnn::operations::data_movement::transpose_hc_multi_core(input_tensor, output_tensor);
         case TransposeOpParallelizationStrategy::MULTI_CORE_CN:
-            return transpose_cn_multi_core(input_tensor, output_tensor);
+            return ttnn::operations::data_movement::transpose_cn_multi_core(input_tensor, output_tensor);
```
Comment (TT-BrianLiu) [inline ttnn/cpp/ttnn/operations/data_movement/transpose/device/transpose_op.cpp:132]: I would move these into `detail::` and rename to `_program_factory`.

File: ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp [hunk 11]
```diff
@@ -6,24 +6,95 @@
...
+    bool pad_c = false;
+    bool pad_n = false;
+
+    switch (transpose_dim) {
+        case TransposeOpDim::HC:
+            pad_c = true;
+            break;
+        case TransposeOpDim::NH:
+            return ttnn::permute((const ttnn::Tensor)a, std::vector<int64_t>({2, 1, 0, 3}), output_mem_config);
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp:27]: why do you need this cast `(const ttnn::Tensor)a` ?

**Review by TT-BrianLiu — APPROVED** [review]
Changes look good if tests pass. Some minor comments.

File: ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp (hunk 9 above)
Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp:18]: this is just a helper function only used within this file.

File: ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp (hunk 11 above)
Comment (tarafdarTT) [reply ttnn/cpp/ttnn/operations/data_movement/transpose/transpose.hpp:27]: I do not, I'll remove

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]

**Review by skhorasganiTT — APPROVED** [review]

**Review by eyonland — APPROVED** [review]
Wish we could get rid of that auto format....

**Review by tt-aho — APPROVED** [review]

**Review by djordje-tt — APPROVED** [review]

**Review by johanna-rock-tt — APPROVED** [review]
falcon40b changes look good to me

**Review by mtairum — APPROVED** [review]

**Review by boris-drazic — APPROVED** [review]

File: tests/tt_eager/python_api_testing/non_working_unit_tests/wormhole/test_dispatch_mode_transpose_nh.py [hunk 12]
```diff
@@ -20,7 +21,7 @@
 
 def tt_transpose_nh(x, device, dtype, layout, input_mem_config, output_mem_config):
     t0 = setup_tt_tensor(x, device, layout[0], input_mem_config[0], dtype[0])
-    t1 = ttl.tensor.transpose(t0, 0, 2, output_mem_config=output_mem_config)
+    t1 = ttl.tensor.transpose(t0, 0, 2, memory_config=output_mem_config)
```
Comment (yieldthought) [inline tests/tt_eager/python_api_testing/non_working_unit_tests/wormhole/test_dispatch_mode_transpose_nh.py:24]: Did you mean `ttnn.transpose` here?

**Review by yieldthought — CHANGES_REQUESTED** [review]
Will this also remove `ttnn.experimental.tensor.transpose`? This is [used in mixtral](https://github.com/tenstorrent/tt-metal/blob/87cafeed8bfe62d0e71287510b0b23d29f3f8523/models/demos/t3000/mixtral8x7b/tt/mixtral_attention.py#L248) for example.

Am I correct in understanding this doesn't delete the old code yet, so there will be a period of deprecation for model devs to update?

File: tests/tt_eager/python_api_testing/non_working_unit_tests/wormhole/test_dispatch_mode_transpose_nh.py (hunk 12 above)
Comment (tarafdarTT) [reply tests/tt_eager/python_api_testing/non_working_unit_tests/wormhole/test_dispatch_mode_transpose_nh.py]: oops ya I changed it , thanks @yieldthought

**Review by yieldthought — APPROVED** [review]

**Review by sjameelTT — APPROVED** [review]

**Pull request merged**